lib/libc/newlib/CMakeLists.txt

```diff
@@ -26,12 +26,10 @@ if(LIBC_LIBRARY_DIR)
   set(LIBC_LIBRARY_DIR_FLAG -L${LIBC_LIBRARY_DIR})
 endif()
 
-# Define _ANSI_SOURCE in order to prevent Newlib from defining POSIX primitives
-# in its headers when GNU dialect is used (-std=gnu*). Newlib features.h
-# defines _DEFAULT_SOURCE when __STRICT_ANSI__ is not defined by GCC (i.e. when
-# -std=gnu*), which leads to various POSIX definitions being provided by the
-# Newlib headers and conflicts with the POSIX definitions provided by Zephyr.
-zephyr_compile_definitions(_ANSI_SOURCE)
+# define _POSIX_C_SOURCE to pull in a few things from the
+# posix definitions that are used like strnlen, and how
+# strerror_r is defined.
+zephyr_compile_options(-D_POSIX_C_SOURCE=200809)
```

Comment:
We definitely do not want to build with `_POSIX_C_SOURCE`, which ends up defining `__POSIX_VISIBLE` and pulling all sorts of POSIX definitions that we do not want.

Until we patch Newlib to support something like `_ZEPHYR_SOURCE` that only exposes the POSIX/non-standard functions that are allowed to be used in the Zephyr codebase (see https://github.com/zephyrproject-rtos/zephyr/pull/57598), we should just locally define the prototypes of the allowed functions in the `lib/libc/newlib`.

Reply:
How does picolibc handle this?  We define it for for picolibc here:

https://github.com/zephyrproject-rtos/picolibc/blob/main/zephyr/zephyr.cmake#L86

Reply:
> How does picolibc handle this? We define it for for picolibc here:

Yea, we need to fix that.

We have already discussed porting Picolibc to "natively" support Zephyr and not expose the POSIX stuff. Until we do that though, we can provide the non-standard function prototypes (e.g. `strnlen`) in `lib/libc/picolibc`.

Reply:
Not sure how to define strerror_r to match the POSIX definition.  If I introduce `lib/libc/newlib/include/string.h` with:

```
extern size_t strnlen(const char *s, size_t maxlen);
extern char    *strerror_r (int, char *, size_t);
extern char  *strtok_r(char *str, const char *sep, char **state);
```

I get the following build error:
```
In file included from /home/galak/git/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:19,
                 from /home/galak/git/zephyr/subsys/testsuite/ztest/include/zephyr/ztest.h:49,
                 from /home/galak/git/zephyr/tests/lib/c_lib/src/test_qsort.c:10:
/home/galak/git/zephyr/lib/libc/newlib/include/string.h:20:14: error: conflicting types for 'strerror_r'; have 'int(int,  char *, size_t)' {aka 'int(int,  char *, unsigned int)'}
   20 | extern int   strerror_r(int errnum, char *strerrbuf, size_t buflen);
      |              ^~~~~~~~~~
In file included from /home/galak/git/zephyr/lib/libc/newlib/include/string.h:13:
/opt/toolchains/zephyr-sdk-0.15.2/xtensa-sample_controller_zephyr-elf/xtensa-sample_controller_zephyr-elf/sys-include/string.h:97:10: note: previous declaration of 'strerror_r' with type 'char *(int,  c
har *, size_t)' {aka 'char *(int,  char *, unsigned int)'}
   97 | char    *strerror_r (int, char *, size_t);
      |          ^~~~~~~~~~
```

Reply:
It looks like `__GNU_VISIBLE` is being defined somewhere?

https://github.com/zephyrproject-rtos/newlib-cygwin/blob/4e150303bcc1e44f4d90f3489a4417433980d5ff/newlib/libc/include/string.h#L96-L97

Reply:
> > How does picolibc handle this? We define it for for picolibc here:
> 
> Yea, we need to fix that.
> 
> We have already discussed porting Picolibc to "natively" support Zephyr and not expose the POSIX stuff. Until we do that though, we can provide the non-standard function prototypes (e.g. `strnlen`) in `lib/libc/picolibc`.

Is there any reason not to respect the existing standards where files needing various extended APIs just define the relevant `_SOURCE` macro before including the headers? If you want POSIX stuff in your application, define the `_POSIX_C_SOURCE` macro, if you want GNU stuff, define `_GNU_SOURCE`. Use existing standards for now, and when you want to make a formal Zephyr API specification, expect C libraries to respect `_ZEPHYR_SOURCE`.

Reply:
> Is there any reason not to respect the existing standards where files needing various extended APIs just define the relevant `_SOURCE` macro before including the headers? If you want POSIX stuff in your application, define the `_POSIX_C_SOURCE` macro, if you want GNU stuff, define `_GNU_SOURCE`.

Mainly because that creates a lot of clutter and also that in itself is not standardised.

Also with https://github.com/zephyrproject-rtos/zephyr/pull/57598 making exemptions for the non-standard functions like `strnlen()`, we are making a promise that these non-standard functions will always be available regardless of the selected libc without requiring any special user interventions (e.g. defining `_POSIX_C_SOURCE`).

Reply:
I'd much rather see this done by (temporarily) defining the relevant C library enabler macros instead of creating potentially conflicting prototypes in zephyr headers. Applications which need to define `_GNU_SOURCE` could easily run into trouble with the approach proposed here as the Zephyr prototypes might not exactly match the C library prototypes.

Fixing picolibc to respect `_ZEPHYR_SOURCE` is not going to be a huge task; let's get started on that instead of adding temporary kludge-arounds in Zephyr itself.

Reply:
> Applications which need to define `_GNU_SOURCE` could easily run into trouble with the approach proposed here as the Zephyr prototypes might not exactly match the C library prototypes.

We are defining the prototypes per libc according to the one provided by the specific libc, so I do not think there should ever be a case where the prototypes do not match?

If you are referring to GNU vs. POSIX, Zephyr codebase should stick with only one.

> Fixing picolibc to respect `_ZEPHYR_SOURCE` is not going to be a huge task; let's get started on that instead of adding temporary kludge-arounds in Zephyr itself.

That would be great.

Reply:
Let's move this discussion to https://github.com/zephyrproject-rtos/zephyr/pull/57619 where this proposal is being implemented.